File: BOJ/1976.여행 가자/solution.java
import java.io.*;
import java.util.StringTokenizer;

public class Main {

    private static int n;
    private static int m;

    private static boolean[] visited;
    private static int[][] graph;
    private static int[] plan;

    public static void main(String[] args) throws IOException, NumberFormatException {
        BufferedReader input = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter output = new BufferedWriter(new OutputStreamWriter(System.out));

        StringTokenizer st;

        boolean answer = true;

        n = Integer.parseInt(input.readLine());
        m = Integer.parseInt(input.readLine());

        visited = new boolean[n];

        graph = new int[n][n];
        plan = new int[m];

        for (int i = 0; i < n; i++) {
            st = new StringTokenizer(input.readLine());
            for (int j = 0; j < n; j++) {
                graph[i][j] = Integer.parseInt(st.nextToken());
            }
        }

        st = new StringTokenizer(input.readLine());
        for (int i = 0; i < m; i++) {
            plan[i] = Integer.parseInt(st.nextToken());
        }

        DFS(plan[0] - 1);

        for (int p : plan) {
            if (p > 0 && !visited[p - 1]) {
                answer = false;
                break;
            }
        }

        output.write(answer ? "YES" : "NO");

        output.flush();
        output.close();
        input.close();
    }

    public static void DFS(int x) {
        visited[x] = true;

        for (int i = 0; i < n; i++) {
            if (graph[x][i] == 1 && !visited[i]) {
                DFS(i);
            }
        }
    }
}

File: BOJ/1976.여행 가자/solution.swift
import Foundation

func DFS(x: Int) {
    visited[x] = true

    for i in 0 ..< n {
        if graph[x][i] == 1 && visited[i] == false {
            DFS(x: i)
        }
    }
}

var answer: Bool = true

var n: Int = Int(readLine()!)!
var m: Int = Int(readLine()!)!

var graph: [[Int]] = []
var plan: [Int] = []
var visited: [Bool] = Array(repeating: false, count: n)

for _ in 0 ..< n {
    graph.append(readLine()!.split(separator: " ").map { Int($0)! })
}

plan = readLine()!.split(separator: " ").map { Int($0)! }

DFS(x: plan[0]-1)

for p in plan {
    if !visited[p-1] {
        answer = false
        break
    }
}

print(answer ? "YES" : "NO")
